main/src/ca/uwaterloo/flix/language/ast/SimplifiedAst.scala

```diff
@@ -730,4 +730,236 @@ object SimplifiedAst {
 
   case class Property(law: Law, exp: SimplifiedAst.Expression, loc: SourceLocation) extends SimplifiedAst
 
+
+  /**
+   * A trait to consume a SimplifiedAst.Expression
+   * and produce a new Expression that's modified
+   * in some way. Typical usage is to extend the
+   * trait and override which methods corresponding
+   * to whatever case of Expression you're interested
+   * in and then call the `foldExpression` method
+   * to begin modifying a given expression.
+   */
+  trait ExpressionFolder {
+
+    /**
+      * This is the entry point for the folder. It recursively visits every Expression
+      * calling out to the overridable fold* methods.
+      *
+      * @param expr The expression to fold over and potentially modify.
+      * @return The potentially modified Expression.
+      */
+    final def foldExpression(expr: Expression): Expression = expr match {
```

Comment:
The return type should probably be generic for the visitor to be useful in general, right?